### PR TITLE
Auth-core-redo

### PR DESCRIPTION
**Pull request by dowski**
Includes a number of changes to adapt forked @auth/core code from 0.37 for our purposes.

Includes contributions from dowski and sshader.

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by sshader — COMMENTED** [review]
I think there's also https://github.com/get-convex/convex-auth/blob/main/src/server/provider_utils.ts

Once we get something working, would love to see this rebased into commits of straight copies + the manual changes we make on top so any future upgrades will be easier

File: test/src/main.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import ReactDOM from "react-dom/client";
 import App from "./App.tsx";
 import "./index.css";
 
-const convex = new ConvexReactClient(import.meta.env.VITE_CONVEX_URL as string);
+const convex = new ConvexReactClient(import.meta.env.VITE_CONVEX_URL as string, {verbose: true});
```
[the remark this answers is not in the document]
Comment (dowski) [reply test/src/main.tsx:9]: Actually I'm gonna leave it. Probably helpful.

**Comment by dowski** [comment]
> I think there's also https://github.com/get-convex/convex-auth/blob/main/src/server/provider_utils.ts
> 
> Once we get something working, would love to see this rebased into commits of straight copies + the manual changes we make on top so any future upgrades will be easier

Okay, the whole thing is rebased on top of a straight import of the forked versions. We could optionally check that commit in separately if we want the pristine versions in our actual commit history. I have a branch with just that change if we want to do that.

File: test/src/main.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (sshader) [reply test/src/main.tsx:9]: I believe this gets mirrored to https://github.com/get-convex/convex-auth-example/ via a script, so can we back this out?

**Review by sshader — APPROVED** [review]
If we've done some manual testing, I'm down.

My last request would be to rebase this cleanly so it's straight copy paste from `@auth/core` followed by all manual changes (and don't squash the commits), but if that's too time consuming down to just merge as is

**Comment by dowski** [comment]
Thanks for the review!

> My last request would be to rebase this cleanly so it's straight copy paste from `@auth/core` followed by all manual changes (and don't squash the commits), but if that's too time consuming down to just merge as is

That's why I did in [84c10c7](https://github.com/get-convex/convex-auth/pull/121/commits/84c10c7a1c3e5ae65fb8b7c8379e9a8747d21edb) - maybe what you mean is that should be checked in separately so it's part of the repo history?

File: test/src/main.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dowski) [reply test/src/main.tsx]: Yep!

**Comment by dowski** [comment]
> > My last request would be to rebase this cleanly so it's straight copy paste from `@auth/core` followed by all manual changes (and don't squash the commits), but if that's too time consuming down to just merge as is
> 
> ... maybe what you mean is that should be checked in separately so it's part of the repo history?

Okay, that part is now in a separate PR (#130).

File: src/server/oauth/providers/oauth.ts [hunk 2]
```diff
@@ -270,7 +26,10 @@ export type OAuthConfigInternal<Profile> = Omit<
     url: URL;
     request?: TokenEndpointHandler["request"];
     clientPrivateKey?: CryptoKey | PrivateKey;
-    /** @internal */
+    /**
+     * @internal
+     * @deprecated
```
Comment (sshader) [inline src/server/oauth/providers/oauth.ts:31]: we probably shouldn't be adding a `@deprecated`? Potentially from a bad rebase?

**Review by sshader — APPROVED** [review]
Neat!

File: src/server/oauth/providers/oauth.ts (hunk 2 above)
Comment (dowski) [reply src/server/oauth/providers/oauth.ts:31]: Oh woops, missed this comment. I'll address in a follow-up.

**Pull request merged**